lib/order.rb

```diff
@@ -8,11 +10,58 @@ def initialize(id, products)
     end
 
     def total
-      # TODO: implement total
+      # returns order total with tax
+      sum = products.values.inject(0, :+)
+      total = (sum + (sum * 0.075)).round(2)
+      return total
     end
 
     def add_product(product_name, product_price)
-      # TODO: implement add_product
+      #  adds a product if product doesn't exist
+      if @products.has_key?(product_name)
+        return false
+      else
+        @products[product_name] = product_price
+        return true
+      end
+    end
+
+    def remove_product(product_name)
+      # deletes product if the product is in the order
+      if @products.has_key?(product_name)
+        @products.delete(product_name)
+        return true
+      else
+        return false
+      end
+    end
+
+    def self.all
+      orders = []
+      CSV.open("support/orders.csv", "r").each do |line|
+        data = []
+        product_data = line[1].split(/:|;/)
+        product_data.each_with_index do |datum, i|
+          if i % 2 == 0
```

Comment:
Good work getting this complex logic working. Another way that I've seen to handle it:

```ruby
CSV.open("support/orders.csv", "r").each do |line|
  parsed_products = {}
  raw_products = line[1].split(';')
  raw_products.each do |product|
    name, price = product.split(':')
    parsed_products[name] = price
  end
  order = new(line[0].to_i, parsed_products)
  orders.push(order)
end
```
This removes the need for the inner `if` statement, but also means you're calling `split` twice. Neither one is "right", but I always find it interesting to compare different approaches like this.